### PR TITLE
Added `d1 export --local` support

### PR DESCRIPTION
**Pull request by geelen**
This uses a new miniflare-only debug pragma `PRAGMA miniflare_d1_export` to trigger the dump as D1 provides no in-Worker API (which Miniflare uses) to generate exports.

I'm open to other ways to implement this.

## What this PR solves / how to test

Fixes limitation with the previous iteration of the export functionality. Previously, all `d1 export --local` commands would fail.

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required / Maybe required
  - [x] Not required because: covered by local tests
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Not necessary because: documentation doesn't say `--local` doesn't work, it just emphasises `--remote` (which still makes sense to be the default).

### CODE REVIEW

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
+// NOTE: this function duplicates the logic inside SQLite's shell.c.in as close
+// as possible, with any deviations noted.
+import { SqlStorage } from "@cloudflare/workers-types/experimental";
+
+export function* dumpSql(
```
Comment (andyjessop) [inline packages/miniflare/src/workers/d1/dumpSql.ts:5]: Is it worth building the array here instead of using a generator? It could simplify the code and would mean it wouldn't need to convert to an array in the caller. Or is there another reason why the sql return value needs to be in iterable?

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+				if (cell === null) {
+					return "NULL";
+				} else if (colType === "INTEGER" || cellType === "number") {
+					return cell;
+				} else if (colType === "TEXT" || cellType === "string") {
+					// Reproduce `output_quoted_escaped_string` from https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L1799-L1862
+					let lfs = false;
+					let crs = false;
+					// First escape any ' as ''
+					let output_string = `'${(cell as string).replace(
```
Comment (andyjessop) [inline packages/miniflare/src/workers/d1/dumpSql.ts:75]: Nit: this could potentially be extracted to a separate function to make the `sqlDump` function more readable.

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    ORDER BY tbl_name='sqlite_sequence', rowid;
+  `)();
+	const tables: any[] = Array.from(tables_cursor);
+
+	for (const { name: table, sql } of tables) {
+		if (filterTables.size > 0 && !filterTables.has(table)) continue;
+
+		if (table === "sqlite_sequence") {
+			if (!noSchema) yield `DELETE FROM sqlite_sequence;`;
+		} else if (table.startsWith(`sqlite_stat`)) {
```
Comment (justin-mp) [inline packages/miniflare/src/workers/d1/dumpSql.ts:35]: This matches too many table names!  It's supposed to only match sqlite_stat plus exactly one more character.   The code in shell.in.c's `dump_callback` is `sqlite3_strglob("sqlite_stat?", zTable)==0`, and the `?` operator in a glob matches exactly one character.  I don't know if JS has globbing, but the equivalent regexp would be `/sqlite_stat./`.

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		tables?: string[];
+	}
+) {
+	yield `PRAGMA defer_foreign_keys=TRUE;`;
+
+	// Empty set means include all tables
+	const filterTables = new Set(options?.tables || []);
+	const { noData, noSchema } = options || {};
+
+	// Taken from SQLite shell.c.in https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L8444-L8450
```
Comment (justin-mp) [inline packages/miniflare/src/workers/d1/dumpSql.ts:19]: Can you put a permalink here?  I just pulled master and this SQL is on line 8464 now.

File: packages/miniflare/src/workers/d1/dumpSql.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+					console.log({ colType, cellType, cell, column: columns[i] });
+					return "ERROR";
+				}
+			});
+
+			yield `INSERT INTO ${sqliteQuote(table)} VALUES(${formattedCells.join(",")});`;
+		}
+	}
+
+	if (!noSchema) {
```
Comment (justin-mp) [inline packages/miniflare/src/workers/d1/dumpSql.ts:108]: Does it matter that this query appears first in shell.c.in?

File: packages/wrangler/src/__tests__/d1/export.test.ts [hunk 6]
```diff
@@ -14,19 +14,90 @@ describe("execute", () => {
...
 			`Missing required argument: output`
 		);
 	});
 
-	it("should reject --local mode (for now)", async () => {
-		await expect(
-			runWrangler("d1 export db --local --output /tmp/test.sql")
-		).rejects.toThrowError(
-			`Local imports/exports will be coming in a future version of Wrangler.`
+	it("should handle local", async () => {
```
Comment (justin-mp) [inline packages/wrangler/src/__tests__/d1/export.test.ts:25]: You should probably add tests for the edge cases like quoted names adding the "IF NOT EXISTS" as well as properly quoting things like keywords.

**Review by justin-mp — COMMENTED** [review]
For my own education, does this PR mean that `PRAGMA miniflare_d1_export(?,?,?);` is part of our public API now?  I don't think we want it to be so.  How do we prevent people from writing that in their own queries?

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 1 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:5]: This is for parity with internal D1 code which uses a generator as it needs to stream the results out with backpressure. For Miniflare we keep the same code (duplicated as it didn't feel worth extracting a shared package for a single file that should very rarely change) but accumulate into an array in the caller.

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 2 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:75]: Yeah I wrote this file while trying my best to copy line-by-line the functionality in shell.c.in, but I think it makes sense to then slightly restructure it for readability after the fact.

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 1 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:5]: Though, in saying that, it would be nice to have a separate repo with a bunch of unit tests that cover regressions. At the moment that's inside D1's internal codebase, so I think I'll add a cleanup ticket to extract this out and unify the references. It would be nice to have this released in the mean time though.

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 3 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:35]: Good catch! Fixed.

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 4 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:19]: 👍

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 5 above)
Comment (geelen) [reply packages/miniflare/src/workers/d1/dumpSql.ts:108]: Does it? This is supposed to match up with https://github.com/sqlite/sqlite/blob/105c20648e1b05839fd0638686b95f2e3998abcb/src/shell.c.in#L8472, i.e. directly after `run_schema_dump_query` (which dumps schema & data), it checks `SHFLG_DumpDataOnly)` is not set before exporting all `('index','trigger','view')` rows of `sqlite_schema`

Have I got that wrong?

File: packages/wrangler/src/__tests__/d1/export.test.ts (hunk 6 above)
Comment (geelen) [reply packages/wrangler/src/__tests__/d1/export.test.ts:25]: I have all these inside the D1 codebase, but see https://github.com/cloudflare/workers-sdk/pull/6073#discussion_r1655712242 I think it makes sense for this file to live in a separate package with a whole suite of tests (as well as a test runner that runs native SQLite for comparison like the D1 suite does).

Imo that's a follow-up task. For the moment I'm just keeping the two implementations in sync manually.

**Comment by geelen** [comment]
> For my own education, does this PR mean that `PRAGMA miniflare_d1_export(?,?,?);` is part of our public API now? I don't think we want it to be so. How do we prevent people from writing that in their own queries?

It's part of miniflare's implementation of D1, but not present anywhere else. It's not accessible from `wrangler d1 execute --local` since you can't pass parameters but technically if someone wrote it in a worker and passed in at least two params they'd get a sql dump in local dev (and an error in production).

We could refactor `dumpSql` in this PR to use the public D1 API rather than the internal DO one (and then `wrangler export` would just send lots of queries to the D1 rather than a single debug pragma), but that would mean it's much harder to reuse the specific `dumpSql` code that's present in D1's internal codebase. And so I've erred on the side of making it easier to keep them consistent and accept leaking an implementation detail.

Ideally, Miniflare would make it possible to do an end-run around the D1 shim and hit an object directly with an RPC call, but as far as I can find out that's not possible.

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 1 above)
Comment (andyjessop) [reply packages/miniflare/src/workers/d1/dumpSql.ts:5]: Makes sense, thank you!

**Review by andyjessop — APPROVED** [review]

File: packages/miniflare/src/workers/d1/dumpSql.ts (hunk 5 above)
Comment (justin-mp) [reply packages/miniflare/src/workers/d1/dumpSql.ts:86]: No, you're correct.

File: packages/wrangler/src/__tests__/d1/export.test.ts (hunk 6 above)
Comment (justin-mp) [reply packages/wrangler/src/__tests__/d1/export.test.ts:25]: I agree it could be a follow-up task.

**Review by justin-mp — APPROVED** [review]
